alg/teca_spectral_filter.py

```diff
@@ -0,0 +1,253 @@
+        The order is typically a ordinal between 1 and 6.  A ValueError is
+        raised when unsupported filter type is encountered """
+
+        if ftype != 'low_pass' and ftype != 'high_pass':
+            raise ValueError('Invalid filter type %s. The filter type may be '
+                             'low_pass or high_pass') % (ftype)
+
+        if order < 1 or order > 6:
+            raise ValueError('The filter order should be between 1 and 6 but '
+                             'is %f' % (order))
+
+        self.filter_type = ftype
+        self.filter_order = order
+
+    def set_critical_period(self, period, units):
+        """ Set the filter's critical frequency from a period given in the
+        specified units. The units may be seconds, minutes, hours or days. This
+        could be used to filter on scales of familiar units. Eg. 3 days """
+
+        self.critical_frequency = 1. / ( period * self.get_number_of_seconds(units) )
+
+    def set_critical_frequency(self, freq):
+        """ Set the filter's critical frequency in units of Hz """
+
+        self.critical_frequency = freq
+
+    def set_pass_input_arrays(self, pass_in):
+        """ When set to True the input arrays will be included in the output.
+        In that case the filtered array will be named: <name>_<filter_type>
+        where name is the name of the input array and filter type is one of
+        low_pass or high_pass """
+
+        self.pass_input_arrays = pass_in
+
+    def set_point_arrays(self, arrays):
+        """ Set the list of arrays to filter """
+
+        if not isinstance(arrays, list):
+            arrays = list(arrays)
+
+        self.point_arrays = arrays
+
+    @staticmethod
+    def get_number_of_seconds(unit):
+        """ get the number of seconds in the specified unit. unit string must
+        contaoin one of seconds, minutes, hours, or days. A ValueError is
+        raised when invalid units are encountered """
+
+        if 'seconds' in unit:
+            return 1.
+        elif 'minutes' in unit:
+            return 60.
+        elif 'hours' in  unit:
+            return 60.*60.
+        elif 'days' in unit:
+            return 60.*60.*24.
+
+        raise ValueError('Unsupported time axis units %s' % (unit))
+
+    @staticmethod
+    def butterworth(npmod, n_samples, sample_rate, critical_freq, order, high_pass):
+        """
+        Generate a Butterworth filter that can be multiplied with a signal in
+        the frequency domain.
+
+        npmod - the compute module (numpy or cupy)
+        n_samples - filter size in number of samples
+        sample_rate - sample rate in Hz
+        critical_freq - cut off frequency in Hz
+        order - filter order
+        high_pass - if true the high pass version is generated
+        """
+        fs2 = sample_rate / 2.
+        f = npmod.linspace(-fs2, fs2, n_samples)
+        if high_pass:
+            h =  1. - 1. / npmod.sqrt( 1. + npmod.power( f / critical_freq, 2*order ) )
```

Comment:
Only the zero-frequency part of the high-pass has 1 added to it.  Because of this, converting from a high-pass to a low-pass filter should take two lines, something like:
```python
h_high = -h_low
h_high[0] += 1
```

Reply:
I took 1 and subtracted the low pass response which then "inverts" the response. Here's a plot for the numpy example you referenced w/ critical frequency 15 hz:
![bw_high_pass](https://user-images.githubusercontent.com/115053/181658906-f4632129-da4f-472a-a905-a78b1c919db4.png)
You can see the low frequencies are filtered while the high frequencies are passed. I think you're saying I should pass the "DC frequency" as well when doing a high pass filter. If so I think that make sense.

Reply:
I think that might work in this specific case, but it can be proved (I have this proof in a notebook somewhere, and I'll upload it if I track it down) that the formulation that I describe above is exactly identical to taking the original signal and subtracting-off the low-pass filter, hence leaving the high-frequency components.

Reply:
OK. Got it. I also found an explicit formula for the Butterworth high pass [here]( https://tttapa.github.io/Pages/Mathematics/Systems-and-Control-Theory/Analog-Filters/Butterworth-Filters.html) and have updated to use that. here's the order 10 from the updated example above:
![butter_kernel_high pass](https://user-images.githubusercontent.com/115053/181775211-3316460c-fd38-424d-8975-05223a657b7b.png)